utxoscanner.go

```diff
@@ -0,0 +1,409 @@
+	GetBlock func(chainhash.Hash, ...QueryOption) (*btcutil.Block, error)
+}
+
+// UtxoScanner batches calls to GetUtxo so that a single scan can search for
+// multiple outpoints. If a scan is in progress when a new element is added, we
+// check whether it can safely be added to the current batch, if not it will be
+// included in the next batch.
+type UtxoScanner struct {
+	started uint32
+	stopped uint32
+
+	cfg *UtxoScannerConfig
+
+	pq        GetUtxoRequestPQ
+	nextBatch []*GetUtxoRequest
+
+	mu sync.Mutex
+	cv *sync.Cond
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// NewUtxoScanner creates a new instance of UtxoScanner using the given chain
+// interface.
+func NewUtxoScanner(cfg *UtxoScannerConfig) *UtxoScanner {
+	scanner := &UtxoScanner{
+		cfg:  cfg,
+		cv:   sync.NewCond(&sync.Mutex{}),
+		quit: make(chan struct{}),
+	}
+	scanner.cv = sync.NewCond(&scanner.mu)
+
+	return scanner
+}
+
+// Start begins running scan batches.
+func (s *UtxoScanner) Start() error {
+	if !atomic.CompareAndSwapUint32(&s.started, 0, 1) {
+		return nil
+	}
+
+	s.wg.Add(1)
+	go s.batchManager()
+
+	return nil
+}
+
+// Stop any in-progress scan.
+func (s *UtxoScanner) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	close(s.quit)
+	s.cv.Signal()
+
+	s.wg.Wait()
+
+	return nil
+}
+
+// Enqueue takes a GetUtxoRequest and adds it to the next applicable batch.
+func (s *UtxoScanner) Enqueue(outPoint *wire.OutPoint,
+	startHeight uint32) (*GetUtxoRequest, error) {
+
+	log.Debugf("Enqueuing request for %s with start height %d",
+		outPoint.String(), startHeight)
+
+	req := &GetUtxoRequest{
+		OutPoint:      outPoint,
+		StartHeight:   startHeight,
+		resultErrChan: make(chan reportAndError, 1),
+		quit:          s.quit,
+	}
+
+	s.cv.L.Lock()
+	select {
+	case <-s.quit:
+		s.cv.L.Unlock()
+		return nil, ErrShuttingDown
+	default:
+	}
+
+	// Insert the request into the queue and signal any threads that might be
+	// waiting for new elements.
+	heap.Push(&s.pq, req)
+
+	s.cv.L.Unlock()
+	s.cv.Signal()
+
+	return req, nil
+}
+
+// batchManager is responsible for scheduling batches of UTXOs to scan. Any
+// incoming requests whose start height has already been passed will be added to
+// the next batch, which gets scheduled after the current batch finishes.
+//
+// NOTE: This method MUST be spawned as a goroutine.
+func (s *UtxoScanner) batchManager() {
+	defer s.wg.Done()
+
+	for {
+		s.cv.L.Lock()
+		// Wait for the queue to be non-empty.
+		for s.pq.IsEmpty() {
+			s.cv.Wait()
+
+			select {
+			case <-s.quit:
+				s.cv.L.Unlock()
+				return
+			default:
+			}
+		}
+
+		// Re-queue previously skipped requests for next batch.
+		for _, request := range s.nextBatch {
+			heap.Push(&s.pq, request)
+		}
+		s.nextBatch = nil
+
+		req := s.pq.Peek()
+		s.cv.L.Unlock()
+
+		// Exit early if a nil request is returned, as this indicates we
+		// are shutting down the utxoscanner.
+		if req == nil {
```

Comment:
Can this really occur? Looks like in the case of shutdown, we'll signal above and return checking the `s.quit` channel. 

The only point at which this returns `nil`, seems to be when the heap is empty, and since we have exited the loop above the heap is non-empty and we are holding the lock.